data/src/main/java/pl/edu/agh/papaya/model/UserInProject.java

```diff
@@ -9,6 +9,7 @@
 import lombok.Getter;
 import lombok.NoArgsConstructor;
 import lombok.Setter;
+import pl.edu.agh.papaya.api.model.UserRole;
```

Comment:
Nie wiem czy to dobry pomysł, żeby w bazie danych trzymać coś, co jest zdefiniowane w api.

Api jest kontraktem między frontem a backendem i powinno być używane tylko do tego. 
Aplikacja powinna być odporna na zmiany api - powinno wystarczyć poprawienie "warstwy" konwertującej modele bazy danych na DTO.

Zgadzam się, że ten enum UserRole nie jest najlepszy, ale jest częścią modelu, a nie częścią DTO.

Reply:
Hm, jest w tym sens, ale trochę głupio, jak to jest tylko na backendzie - chcemy umożliwiać zmianę uprawnień użytkownika

Można zrobić enuma w api, a na backendzie innego enuma mapującego wartości z api w jakieś takie analogiczne wartości w modelu

Reply:
Raczej odwrotnie. Enum powinien być w modelu, a yaml powinien go odwzorowywać.
Wtedy UserRole to np. Admin, User, a w yamlu może być jakieś UserStatus, którego wartością może być też to nieszczęsne `deleted`

Reply:
Poza tym, front nie powinien wiedzieć, że usuwanie użytkownika to zmiana statusu na deleted. Powinno pójść normalne żądanie DELETE z id usera, a ustawieniem flagi deleted zajmuje się backend.

Reply:
> Poza tym, front nie powinien wiedzieć, że usuwanie użytkownika to zmiana statusu na deleted. Powinno pójść normalne żądanie DELETE z id usera, a ustawieniem flagi deleted zajmuje się backend.

:+1: 

> Raczej odwrotnie. Enum powinien być w modelu, a yaml powinien go odwzorowywać.
Wtedy UserRole to np. Admin, User, a w yamlu może być jakieś UserStatus, którego wartością może być też to nieszczęsne deleted

W sensie chcemy się odwoływać w yamlu do enuma zadeklarowanego na backendzie?

Reply:
Tak. Aplikacja ma ustalone grupy uprawnień i one są święte. Front może (ale nie musi) korzystać z identycznych nazw - równie dobrze wystarczyć mu może flaga isAdmin. Nie musi mieć też podanych jako enum wszystkich wartości UserRole.

Reply:
Ja myślałam, żeby w yamlu zrobić enuma `UserStatus`, a na backendzie w enumie `UserRole` byłoby pole z odpowiadającymi wartościami `UserStatus`

> Tak. Aplikacja ma ustalone grupy uprawnień i one są święte. Front może (ale nie musi) korzystać z identycznych nazw - równie dobrze wystarczyć mu może flaga isAdmin. Nie musi mieć też podanych jako enum wszystkich wartości UserRole.

Jeśli chcemy mieć tylko dwa typy uprawnień, to `isAdmin` jest okej. Jakby miało być więcej, to trochę kiepsko

Reply:
Z isAdmin to tylko przykład. Chodziło mi o to, że uprawnienia są ustalone na backendzie i tam powinny pozostać, a yaml jest takim adapterem na to, czego oczekuje front.

Innymi słowy, raczej powinniśmy myśleć o tym odwrotnie - na backendzie są role, a w yamlu odpowiadające im wartości enuma (który może być identyczny, mieć zmienione nazwy, albo dodatkowe wartości np. deleted, którego nie ma w enumie na backendzie).

Reply:
Okeej. Czyli mapowanie z jednego na drugie chcemy robić na podstawie samych stringów w obu enumach? Bo nie wiem, jak inaczej ma działać odwoływanie się w yamlu do de facto kodu backendu

Reply:
W modelu jest UserRole i on też jest wykorzystywany w bazie danych. 
W api jest enum, który zostanie wygenerowany automatycznie (jakiś powiedzmy UserStatusEnum i on będzie w klasie UserDTO).

W metodzie getUsers będzie konwersja z tego co w bazie, na format wymagany przez front (zapewne List<UserDTO>).  Wtedy też będzie konwersja z UserRole na UserStatusEnum (niekoniecznie 1:1, bo może być generowany na podstawie większej liczby źródeł niż tylko UserRole konkretnego usera).

Yaml nie odwołuje się do kodu backendu. Dlatego nie musi mieć identycznych pól/nazw/czegokolwiek jak backend. Dlatego też backend nie może do bazy wstawiać czegoś, co jest generowane z yamla - warstwa modelu i warstwa interfejsu są odseparowane (zmiana api nie może pociągać za sobą zmian w sposobie persystencji/schemacie BD/czymkolwiek innym co nie jest rest controllerem)

Reply:
No wiem, kompletnie nie mówię, żeby wstawiać do bazy enuma z api. Chodzi mi głównie o tę konwersję w `getUsers()` i o to też pytam, bo mówisz, że trzymanie w `UserRole` jako pole listy  odpowiadających danej roli statusów z `UserStatusEnum` z api jest brzydkie. Chcemy serii ifów w `getUsers()` z enigmatyczną logiką?

Reply:
Nie no, będzie jakiś UserRoleConverter wywołany w kontrolerze (spring chyba nawet ma wbudowane konwertery, trzeba by poczytać jak to działa) i on się tym zajmie. W ostateczności switch na enumach chyba nie będzie jakiś bardzo brzydki.

Trzymanie UserRole w modelu i identycznego UserRole w api (o identycznej nazwie/polach) nie jest brzydkie, o ile będą to dwa osobne, niezależne od siebie byty (w takim sensie, że zmiany w jednym - z naciskiem na ten w api - nie powodują konieczności zmian w tym drugim).

Jasne, że jak dodamy nową rolę do aplikacji, to prawdopodobnie front kiedyś też będzie się chciał o tym dowiedzieć. No ale to jest dość duża zmiana - nowa funkcjonalność w pewnym sensie. Ale zmiana w bazie nazwy ADMIN na OWNER nie wyjdzie poza backend - kontrakt z frontem nie zostanie naruszony.

Reply:
Czekaj, jak zmiana w api ma nie wymuszać zmiany na backendzie? Pojawi się jakiś status `MAINTAINER` w api i z frontu będą lecieć przyjazne sugestie, żeby użytkownik Jan Kowalski dostał taki status w projekcie XYZ - to chyba by było fajnie, jakby backend tak jednak musiał się nauczyć, co to oznacza?

Reply:
dlaczego miałby się pojawić nowy status najpierw w api? 
Decyzja "dodajemy nowy typ użytkownika o nazwie maintainer" nie leży po stronie frontu. 
Poza tym (jak wspomniałem wyżej) dodanie nowej roli to duża zmiana, nawet nazwałbym to nową funkcjonalnością (bo to nie tylko dodanie kolejnego `case`, ale również uwzględnienie tego w wielu innych miejscach związanych z autoryzacją.

Reply:
Nie mówię, że to jest decyzja api, tylko "jedno nie ma wymuszać drugiego" zabrzmiało mi podejrzanie (ale tak, wiem, chodzi o dane w bazie)